File: CPP/Array/Quick_Sort.cpp
#include<bits/stdc++.h>

using namespace std;

void print(vector<int> &a){
    for(int i=0;i<a.size();i++){
        cout<<a[i]<<endl;
    }
}
int partition(vector<int> &a, int l, int h){
    int i=l-1, pivot = a[h];
    for(int j=l;j<h;j++){
        if(a[j]<=pivot){
            swap(a[++i],a[j]);
        }
    }
    swap(a[h],a[++i]);
    return i;
}

void quick_sort(vector<int> &a, int l, int h){
    if(l<h){
        int p=partition(a, l, h);
        quick_sort(a, l, p-1);
        quick_sort(a, p+1, h);
    }
}


int main(){
    int n;
    cin>>n;
    vector<int> a(n);
    for(int i=0;i<n;i++){
        cin>>a[i];
    }
    int l=0, h=n-1;

    quick_sort(a, l, h);    
    print(a);
    return 0;
}

File: Java/Sorting/quick_sort.java
// Time Complexity
// Best	O(n*log n)
// Worst	O(n2)
// Average	O(n*log n)
// Space Complexity	O(log n)

import java.util.*;
class quick_sort {
    public static void swap(int[] a, int i, int j){
        int t=a[i];
        a[i]=a[j];
        a[j]=t;
    }
    public static int partition(int[] a, int l, int h){
        int pivot=a[h];
        int i=-1;
        for(int j=0;j<h;j++){
            if(a[j]<pivot){
                i++;
                swap(a, i, j);
            }
        }
        swap(a,i+1,h);
        return i+1;
    }
    public static void sort(int[] a, int l, int h){
        if(l<h){
            int k=partition(a, l, h);
            sort(a, l, k-1);
            sort(a, k+1, h);
        }
    }
    public static void main(String args[]){
        Scanner in=new Scanner(System.in);
        int n=in.nextInt();
        int[] a=new int[n];
        for(int i=0;i<n;i++)
            a[i]=in.nextInt();
        sort(a, 0, n-1);
        for(int i=0;i<n;i++)
            System.out.print(a[i]+" ");
        in.close();
    }
}
